apps/emqx_rule_engine/src/emqx_rule_actions.erl

```diff
@@ -232,3 +243,70 @@ format_pub_props(UserPropertiesTks, Selected, Env) ->
                 replace_simple_var(UserPropertiesTks, Selected, #{})
         end,
     #{'User-Property' => UserProperties}.
+
+format_mqtt_properties(MQTTPropertiesTemplate, Selected, Env) ->
+    #{metadata := #{rule_id := RuleId}} = Env,
+    MQTTProperties0 =
+        maps:fold(
+            fun(K, Template, Acc) ->
+                try
+                    V = emqx_placeholder:proc_tmpl(Template, Selected),
+                    Acc#{K => V}
+                catch
+                    Kind:Error ->
```

Comment:
would be good if `emqx_placeholder` can follow the `throw` pattern for more readable error logs.

Reply:
alternatively, we could return `undefined` when we cannot dive deeper into the structure.  :thinking:

Reply:
I think if value is `undefined`, it's most certainly should be logged sooner than later (as bad_integer later).